ravenml/data/write_dataset.py

```diff
@@ -0,0 +1,235 @@
+import os, shutil, time, json
+import contextlib2
+import tensorflow as tf
+from pathlib import Path
+from random import shuffle
+from datetime import datetime
+from ravenml.data.interfaces import CreateInput
+from ravenml.utils.question import cli_spinner, cli_spinner_wrapper
+from ravenml.utils.config import get_config
+from ravenml.data.helpers import default_filter, default_load_image_ids, copy_data_locally
+
+class DatasetWriter(object):
+
+    def __init__(self, create: CreateInput, **kwargs):
+        self.associated_files = kwargs['associated_files']
+        self.related_data_prefixes = kwargs['related_data_prefixes']
+        
+        metadata = create.metadata
+        self.num_folds = create.kfolds # Never actually used
+        self.test_percent = create.test_percent
+        self.label_to_int_dict = {}
+        self.image_ids = None
+        self.filter_metadata = None
+        self.temp_dir = create.plugin_metadata['temp_dir_path']
+        self.dataset_path = create.dataset_path
+        self.dataset_name = metadata['dataset_name']
+        self.created_by = metadata['created_by']
+        self.comments = metadata['comments']
+        self.plugin_name = create.plugin_metadata['architecture']
+        self.imageset_paths = create.imageset_paths
+        self.filter = metadata['filter']
+        self.cache = create.plugin_cache
+    
+    def construct_all(self, *args, **kwargs):
+        raise NotImplementedError
+
+    def export_data(self, *args, **kwargs):
+        raise NotImplementedError
+    
+    def write_additional_files(self, *args, **kwargs):
+        raise NotImplementedError
+
+    def filter_sets(self, metadata_prefix):
+        tags_df = default_load_image_ids(metadata_prefix, self.imageset_paths)
+        filter_metadata = {"groups": []}
+        if self.filter:
+            self.image_ids = default_filter(tags_df, filter_metadata)
+        else: 
+            self.image_ids = tags_df.index.tolist()
+        
+        self.filter_metadata = filter_metadata
+
+    def load_data(self, metadata_prefix):         
+        if not self.image_ids:
+            tags_df = default_load_image_ids(metadata_prefix, self.imageset_paths)
+            self.image_ids = tags_df.index.tolist()
+            self.filter_metadata = {"groups": []}
```

Comment:
It's a bit confusing and inflexible that loading the image IDs can happen in either of these methods. It should be split out into its own method so that plugins can selectively override it.